cachi2/core/package_managers/generic/main.py

```diff
@@ -47,8 +54,42 @@ def _resolve_generic_lockfile(source_dir: RootedPath, output_dir: RootedPath) ->
 
     log.info(f"Reading generic lockfile: {lockfile_path}")
     lockfile = _load_lockfile(lockfile_path)
+    to_download: dict[str, Union[str, os.PathLike[str]]] = {}
+
+    # output_dir is now the root and cannot be escaped
+    output_dir = output_dir.re_root(DEFAULT_DEPS_DIR)
+
     for artifact in lockfile.artifacts:
         log.debug(f"Resolving artifact: {artifact.download_url}")
+        # make sure target does not point outside the output directory
+        try:
+            destination = output_dir.join_within_root(artifact.resolved_target)
+            artifact._resolved_path = os.path.realpath(destination)
+        except PathOutsideRoot as e:
+            e.solution = (
+                f"Artifact '{artifact.download_url}' has target which falls "
+                "outside of the dependencies root. Refusing to proceed."
+            )
+            raise
+
+        # make sure that targets do not overlap
+        if artifact._resolved_path in to_download.values():
+            raise PackageRejected(
+                f"More than one file from {lockfile_path} is trying to write to the same target: {destination}",
+                solution="Make sure that all targets are unique.",
+            )
+
+        # create the parent directory and assign the resolved path to the artifact
+        Path.mkdir(destination.path.parent, parents=True, exist_ok=True)
+        to_download[str(artifact.download_url)] = artifact._resolved_path
+
+    # disable mypy check to avoid type error on resolved_path being able to be None,
+    # because it is always set in the loop above. Same for checksum verification.
+    asyncio.run(async_download_files(to_download, get_config().concurrency_limit))
+
+    # verify checksums
+    for artifact in lockfile.artifacts:
+        must_match_any_checksum(artifact._resolved_path, artifact.formatted_checksums)
```

Comment:
It would, but I know that these lockfiles won't be small and could go up to a couple thousand. I think that having a little more complicated code to achieve parallel download is worth it in this case.

Reply:
We do parallel pretty much everywhere exactly due to reasons @kosciCZ mentions, so it's preferable and there's precedent -> marking as resolved.

Reply:
ACK